pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java

```diff
@@ -0,0 +1,266 @@
+    _processedRecords.inc();
+  }
+
+  /**
+   * Creates one Pinot segment using the {@link GenericRow}s collected in the AVRO file buffer, at
+   * the outputDirUri as specified in the tableConfig->batchConfigs. Successful invocation of this
+   * method means that the {@link GenericRow}s collected so far, are now available in the Pinot
+   * segment and not available in the buffer anymore.
+   *
+   * <p>Successful completion of segment will return the segment URI, and the URI includes a
+   * sequence id indicating the part number. The sequence id is initialized to 0 and each successful
+   * flush will increment the sequence id by 1. The buffer will be reset and ready to accept further
+   * records via <code>collect()</code> If an exception is thrown, the buffer will not be reset and
+   * so, <code>flush()</code> can be invoked repeatedly in a retry loop. If a successful invocation
+   * is not achieved,<code>close()</code> followed by <code>init </code> will have to be called in
+   * order to reset the buffer and resume record writing.
+   *
+   * @return URI of the generated segment
+   * @throws IOException
+   */
+  @Override
+  public URI flush()
+      throws IOException {
+
+    LOGGER.info("Beginning flush for Pinot table: {} with {} records", _tableNameWithType, _rowCount);
+    _recordWriter.close();
+
+    // Create temp dir for flush
+    File flushDir = new File(_stagingDir, "flush_dir_" + System.currentTimeMillis());
+    Preconditions.checkState(flushDir.mkdirs(), "Failed to create flush dir: %s", flushDir);
+
+    try {
+      // Segment dir
+      File segmentDir = new File(flushDir, "segment_dir");
+
+      // Make BatchIngestionConfig for flush
+      Map<String, String> batchConfigMapOverride = new HashMap<>(_batchConfig.getBatchConfigMap());
+      batchConfigMapOverride.put(BatchConfigProperties.INPUT_DIR_URI, _bufferFile.getAbsolutePath());
+      batchConfigMapOverride.put(BatchConfigProperties.OUTPUT_DIR_URI, segmentDir.getAbsolutePath());
+      batchConfigMapOverride.put(BatchConfigProperties.INPUT_FORMAT, BUFFER_FILE_FORMAT.toString());
+      BatchIngestionConfig batchIngestionConfig = new BatchIngestionConfig(Lists.newArrayList(batchConfigMapOverride),
+          _batchIngestionConfig.getSegmentIngestionType(), _batchIngestionConfig.getSegmentIngestionFrequency());
+
+      // Build segment
+      SegmentGeneratorConfig segmentGeneratorConfig =
+          IngestionUtils.generateSegmentGeneratorConfig(_tableConfig, _schema, batchIngestionConfig);
+      segmentGeneratorConfig.setSequenceId(_seqId);
+      String segmentName = IngestionUtils.buildSegment(segmentGeneratorConfig);
+      LOGGER.info("Successfully built segment: {} of sequence {} for Pinot table: {}", segmentName, _seqId,
+          _tableNameWithType);
+
+      // Tar segment
+      File segmentTarFile =
+          new File(_outputDirURI, String.format("%s_%d%s", segmentName, _indexOfSubtask, Constants.TAR_GZ_FILE_EXT));
```

Comment:
since you added it to the segmentSuffix in BatchConfig, it will already be used by SegmentNameGenerator. Why do we need to explicitly add it in this tar name again?

Reply:
do you mean `_indexOfSubtask`? that's different from the sequence id, `_indexOfSubtask` is the id of the Flink task.

Reply:
yea, indexOfSubtask is added to postfix above
```
String segmentNamePostfixProp = String.format("%s.%s", BatchConfigProperties.SEGMENT_NAME_GENERATOR_PROP_PREFIX,
        BatchConfigProperties.SEGMENT_NAME_POSTFIX);
    String segmentSuffix = batchConfigMap.get(segmentNamePostfixProp);
    segmentSuffix = segmentSuffix == null ? String.valueOf(_indexOfSubtask) : segmentSuffix + "_" + _indexOfSubtask;
    batchConfigMap.put(segmentNamePostfixProp, segmentSuffix);
```